supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala

```diff
@@ -0,0 +1,137 @@
+
+// functionality extracted into the trait for testing purposes
+protected trait PipelinesApiReferenceFilesMappingOperations {
+  private val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  case class ReferenceFile(path: String, crc32c: Long)
+  case class ManifestFile(imageIdentifier: String, diskSizeGb: Int, files: List[ReferenceFile])
+
+  def generateReferenceFilesMapping(auth: GoogleAuthMode,
+                                    referenceDiskLocalizationManifestFiles: Option[List[ValidFullGcsPath]]): PipelinesApiReferenceFilesMapping = {
+
+    val validReferenceFilesMapIO = referenceDiskLocalizationManifestFiles match {
+      case None =>
+        IO.pure(Map.empty[String, PipelinesApiReferenceFilesDisk])
+      case Some(manifestFilesPaths) =>
+        val gcsClient = StorageOptions
+          .newBuilder()
+          .setCredentials(auth.credentials(Set(StorageScopes.DEVSTORAGE_READ_ONLY)))
+          .build
+          .getService
+        val manifestFilesIo = manifestFilesPaths traverse { manifestFilePath => readReferenceDiskManifestFileFromGCS(gcsClient, manifestFilePath) }
+        manifestFilesIo flatMap { manifestFiles =>
+          manifestFiles
+            .traverse(manifestFile => getMapOfValidReferenceFilePathsToDisks(gcsClient, manifestFile))
+            .map(_.flatten.toMap)
+        }
+    }
+    PipelinesApiReferenceFilesMapping(validReferenceFilesMapIO.unsafeRunSync())
+  }
+
+  def getReferenceDisksToMount(pipelinesApiReferenceFilesMapping: PipelinesApiReferenceFilesMapping,
+                               inputFilePaths: Set[String]): List[PipelinesApiReferenceFilesDisk] = {
+    pipelinesApiReferenceFilesMapping.validReferenceFilesMap.filterKeys(key => inputFilePaths.contains(s"gs://$key")).values.toList.distinct
+  }
+
+  protected def readReferenceDiskManifestFileFromGCS(gcsClient: Storage, gcsPath: ValidFullGcsPath): IO[ManifestFile] = {
+    val manifestFileBlobIo = IO { gcsClient.get(BlobId.of(gcsPath.bucket, gcsPath.path.substring(1))) }
+    manifestFileBlobIo flatMap { manifestFileBlob =>
+      val jsonStringIo = IO { manifestFileBlob.getContent().map(_.toChar).mkString }
+      jsonStringIo.flatMap(jsonStr => IO.fromEither(decode[ManifestFile](jsonStr)))
+    }
+  }
+
+  private def getReferenceFileToValidatedGcsPathMap(referenceFiles: Set[ReferenceFile]): IO[Map[ReferenceFile, ValidFullGcsPath]] = {
+    val filesAndValidatedPaths = referenceFiles.map {
+      referenceFile => (referenceFile, GcsPathBuilder.validateGcsPath(s"gs://${referenceFile.path}"))
+    }.toMap
+
+    val (filesWithValidPaths, filesWithInvalidPaths) = filesAndValidatedPaths.partition(_._2.isInstanceOf[ValidFullGcsPath])
+
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new InvalidGcsPathsInManifestFileException(filesWithInvalidPaths.keySet.map(_.path).toList))
+    } else {
+      IO.pure(filesWithValidPaths.mapValues(_.asInstanceOf[ValidFullGcsPath]))
```

Comment:
IO.pure followed by something which might throw an exception makes me nervous - but I think doing the well-typed-lists fix above should remove that nervousness.